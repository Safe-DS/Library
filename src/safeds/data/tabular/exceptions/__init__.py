"""Custom exceptions that can be raised when working with tabular data."""

from ._exceptions import (
    ColumnLengthMismatchError,
    ColumnSizeError,
    DuplicateColumnNameError,
    IndexOutOfBoundsError,
    MissingDataError,
    MissingSchemaError,
    NonNumericColumnError,
    SchemaMismatchError,
    TransformerNotFittedError,
    UnknownColumnNameError,
)

__all__ = [
    "ColumnLengthMismatchError",
    "ColumnSizeError",
    "DuplicateColumnNameError",
    "IndexOutOfBoundsError",
    "MissingDataError",
    "MissingSchemaError",
    "NonNumericColumnError",
    "SchemaMismatchError",
    "TransformerNotFittedError",
    "UnknownColumnNameError",
]
