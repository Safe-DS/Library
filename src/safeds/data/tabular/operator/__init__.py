"""Classes that can store tabular data."""

from typing import TYPE_CHECKING

import apipkg

if TYPE_CHECKING:
    from ._temporal import Temporal

apipkg.initpkg(
    __name__,
    {
        "Temporal": "._temporal:Temporal",
    },
)

__all__ = [
    "Temporal",
]
