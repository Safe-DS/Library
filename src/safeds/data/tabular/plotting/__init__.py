"""Classes that can store tabular data."""

from typing import TYPE_CHECKING

import apipkg

if TYPE_CHECKING:
    from ._experimental_column_plotter import ExperimentalColumnPlotter
    from ._experimental_table_plotter import ExperimentalTablePlotter

apipkg.initpkg(
    __name__,
    {
        "ExperimentalColumnPlotter": "._experimental_column_plotter:ExperimentalColumnPlotter",
        "ExperimentalTablePlotter": "._experimental_table_plotter:ExperimentalTablePlotter",
    },
)

__all__ = [
    "ExperimentalColumnPlotter",
    "ExperimentalTablePlotter",
]
