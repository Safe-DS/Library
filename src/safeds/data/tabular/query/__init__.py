"""Classes that represent queries on the data."""

from typing import TYPE_CHECKING

import apipkg

if TYPE_CHECKING:
    from ._datetime_operations import DatetimeOperations
    from ._duration_operations import DurationOperations
    from ._math_operations import MathOperations
    from ._string_operations import StringOperations

apipkg.initpkg(
    __name__,
    {
        "DatetimeOperations": "._datetime_operations:DatetimeOperations",
        "DurationOperations": "._duration_operations:DurationOperations",
        "MathOperations": "._math_operations:MathOperations",
        "StringOperations": "._string_operations:StringOperations",
    },
)

__all__ = [
    "DatetimeOperations",
    "DurationOperations",
    "MathOperations",
    "StringOperations",
]
