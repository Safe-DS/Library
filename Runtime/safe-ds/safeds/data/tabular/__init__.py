from ._column import Column, ColumnStatistics
from ._row import Row
from ._table import Table
