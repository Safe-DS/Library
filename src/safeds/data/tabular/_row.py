import typing
from typing import Any

import pandas as pd
from IPython.core.display_functions import DisplayHandle, display
from safeds.data.tabular.typing import ColumnType, TableSchema
from safeds.exceptions import UnknownColumnNameError


class Row:
    def __init__(self, data: typing.Iterable, schema: TableSchema):
        self._data: pd.Series = data if isinstance(data, pd.Series) else pd.Series(data)
        self.schema: TableSchema = schema
        self._data = self._data.reset_index(drop=True)

    def __getitem__(self, column_name: str) -> Any:
        return self.get_value(column_name)

    def __iter__(self) -> typing.Iterator[Any]:
        return iter(self.get_column_names())

    def __len__(self) -> int:
        return len(self._data)

    def get_value(self, column_name: str) -> Any:
        """
        Return the value of a specified column.

        Parameters
        ----------
        column_name : str
            The column name.

        Returns
        -------
        value :
            The value of the column.
        """
        if not self.schema.has_column(column_name):
            raise UnknownColumnNameError([column_name])
        return self._data[self.schema._get_column_index_by_name(column_name)]

    def count(self) -> int:
        """
        Return the number of columns in this row.

        Returns
        -------
        count : int
            The number of columns.
        """
        return len(self._data)

    def has_column(self, column_name: str) -> bool:
        """
        Return whether the row contains a given column.
        Alias for self.schema.hasColumn(column_name: str) -> bool.

        Parameters
        ----------
        column_name : str
            The name of the column.

        Returns
        -------
        contains : bool
            True, if row contains the column.
        """
        return self.schema.has_column(column_name)

    def get_column_names(self) -> list[str]:
        """
        Return a list of all column names saved in this schema.
        Alias for self.schema.get_column_names() -> list[str].

        Returns
        -------
        column_names : list[str]
            The column names.
        """
        return self.schema.get_column_names()

    def get_type_of_column(self, column_name: str) -> ColumnType:
        """
        Return the type of a specified column.
        Alias for self.schema.get_type_of_column(column_name: str) -> ColumnType.

        Parameters
        ----------
        column_name : str
            The name of the column.

        Returns
        -------
        type : ColumnType
            The type of the column.

        Raises
        ------
        ColumnNameError
            If the specified target column name does not exist.
        """
        return self.schema.get_type_of_column(column_name)

    def __eq__(self, other: typing.Any) -> bool:
        if not isinstance(other, Row):
            return NotImplemented
        if self is other:
            return True
        return self._data.equals(other._data) and self.schema == other.schema

    def __hash__(self) -> int:
        return hash(self._data)

    def __str__(self) -> str:
        tmp = self._data.to_frame().T
        tmp.columns = self.get_column_names()
        return tmp.__str__()

    def __repr__(self) -> str:
        tmp = self._data.to_frame().T
        tmp.columns = self.get_column_names()
        return tmp.__repr__()

    def _ipython_display_(self) -> DisplayHandle:
        """
        Return a display object for the column to be used in Jupyter Notebooks.

        Returns
        -------
        output : DisplayHandle
            Output object.
        """
        tmp = self._data.to_frame().T
        tmp.columns = self.get_column_names()

        with pd.option_context(
            "display.max_rows", tmp.shape[0], "display.max_columns", tmp.shape[1]
        ):
            return display(tmp)
