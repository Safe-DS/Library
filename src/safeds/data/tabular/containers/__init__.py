from ._column import Column
from ._row import Row
from ._table import Table
from ._tagged_table import TaggedTable
