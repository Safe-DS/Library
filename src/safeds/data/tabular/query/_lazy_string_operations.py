from __future__ import annotations

from typing import TYPE_CHECKING

from safeds._utils import _structural_hash
from safeds._validation import _check_bounds, _ClosedBound, _convert_and_check_datetime_format
from safeds.data.tabular.containers._cell import _to_polars_expression
from safeds.data.tabular.containers._lazy_cell import _LazyCell
from safeds.data.tabular.typing import ColumnType

from ._string_operations import StringOperations

if TYPE_CHECKING:
    import datetime

    import polars as pl

    from safeds._typing import _ConvertibleToIntCell, _ConvertibleToStringCell
    from safeds.data.tabular.containers._cell import Cell


class _LazyStringOperations(StringOperations):
    # ------------------------------------------------------------------------------------------------------------------
    # Dunder methods
    # ------------------------------------------------------------------------------------------------------------------

    def __init__(self, expression: pl.Expr) -> None:
        self._expression: pl.Expr = expression

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, _LazyStringOperations):
            return NotImplemented
        if self is other:
            return True
        return self._expression.meta.eq(other._expression)

    def __hash__(self) -> int:
        return _structural_hash(self._expression.meta.serialize())

    def __repr__(self) -> str:
        return f"_LazyStringOperations({self._expression})"

    def __sizeof__(self) -> int:
        return self._expression.__sizeof__()

    def __str__(self) -> str:
        return f"({self._expression}).str"

    # ------------------------------------------------------------------------------------------------------------------
    # String operations
    # ------------------------------------------------------------------------------------------------------------------

    def contains(self, substring: _ConvertibleToStringCell) -> Cell[bool | None]:
        substring = _to_polars_expression(substring, type_if_none=ColumnType.string())

        return _LazyCell(self._expression.str.contains(substring, literal=True))

    def ends_with(self, suffix: _ConvertibleToStringCell) -> Cell[bool | None]:
        suffix = _to_polars_expression(suffix)

        return _LazyCell(self._expression.str.ends_with(suffix))

    def index_of(self, substring: _ConvertibleToStringCell) -> Cell[int | None]:
        substring = _to_polars_expression(substring, type_if_none=ColumnType.string())

        return _LazyCell(self._expression.str.find(substring, literal=True))

    def length(self, optimize_for_ascii: bool = False) -> Cell[int | None]:
        if optimize_for_ascii:
            return _LazyCell(self._expression.str.len_bytes())
        else:
            return _LazyCell(self._expression.str.len_chars())

    def pad_end(self, length: int, *, character: str = " ") -> Cell[str | None]:
        _check_bounds("length", length, lower_bound=_ClosedBound(0))
        if len(character) != 1:
            raise ValueError("Can only pad with a single character.")

        return _LazyCell(self._expression.str.pad_end(length, character))

    def pad_start(self, length: int, *, character: str = " ") -> Cell[str | None]:
        _check_bounds("length", length, lower_bound=_ClosedBound(0))
        if len(character) != 1:
            raise ValueError("Can only pad with a single character.")

        return _LazyCell(self._expression.str.pad_start(length, character))

    def repeat(self, count: _ConvertibleToIntCell) -> Cell[str | None]:
        if isinstance(count, int):
            _check_bounds("count", count, lower_bound=_ClosedBound(0))

        count = _to_polars_expression(count)

        return _LazyCell(self._expression.repeat_by(count).list.join("", ignore_nulls=False))

    def remove_prefix(self, prefix: _ConvertibleToStringCell) -> Cell[str | None]:
        prefix = _to_polars_expression(prefix, type_if_none=ColumnType.string())

        return _LazyCell(self._expression.str.strip_prefix(prefix))

    def remove_suffix(self, suffix: _ConvertibleToStringCell) -> Cell[str | None]:
        suffix = _to_polars_expression(suffix, type_if_none=ColumnType.string())

        return _LazyCell(self._expression.str.strip_suffix(suffix))

    def replace_all(self, old: _ConvertibleToStringCell, new: _ConvertibleToStringCell) -> Cell[str | None]:
        old = _to_polars_expression(old, type_if_none=ColumnType.string())
        new = _to_polars_expression(new, type_if_none=ColumnType.string())

        return _LazyCell(self._expression.str.replace_all(old, new, literal=True))

    def reverse(self) -> Cell[str | None]:
        return _LazyCell(self._expression.str.reverse())

    def slice(
        self,
        *,
        start: _ConvertibleToIntCell = 0,
        length: _ConvertibleToIntCell = None,
    ) -> Cell[str | None]:
        if isinstance(length, int):
            _check_bounds("length", length, lower_bound=_ClosedBound(0))

        start = _to_polars_expression(start)
        length = _to_polars_expression(length)

        return _LazyCell(self._expression.str.slice(start, length))

    def starts_with(self, prefix: _ConvertibleToStringCell) -> Cell[bool | None]:
        prefix = _to_polars_expression(prefix)

        return _LazyCell(self._expression.str.starts_with(prefix))

    def strip(self, *, characters: _ConvertibleToStringCell = None) -> Cell[str | None]:
        characters = _to_polars_expression(characters)

        return _LazyCell(self._expression.str.strip_chars(characters))

    def strip_end(self, *, characters: _ConvertibleToStringCell = None) -> Cell[str | None]:
        characters = _to_polars_expression(characters)

        return _LazyCell(self._expression.str.strip_chars_end(characters))

    def strip_start(self, *, characters: _ConvertibleToStringCell = None) -> Cell[str | None]:
        characters = _to_polars_expression(characters)

        return _LazyCell(self._expression.str.strip_chars_start(characters))

    def to_date(self, *, format: str | None = "iso") -> Cell[datetime.date | None]:
        if format == "iso":
            format = "%F"  # noqa: A001
        elif format is not None:
            format = _convert_and_check_datetime_format(format, type_="date", used_for_parsing=True)  # noqa: A001

        return _LazyCell(self._expression.str.to_date(format=format, strict=False))

    def to_datetime(self, *, format: str | None = "iso") -> Cell[datetime.datetime | None]:
        if format == "iso":
            format = "%+"  # noqa: A001
        elif format is not None:
            format = _convert_and_check_datetime_format(format, type_="datetime", used_for_parsing=True)  # noqa: A001

        return _LazyCell(self._expression.str.to_datetime(format=format, strict=False))

    def to_float(self) -> Cell[float | None]:
        import polars as pl

        return _LazyCell(self._expression.cast(pl.Float64(), strict=False))

    def to_int(self, *, base: _ConvertibleToIntCell = 10) -> Cell[int | None]:
        base = _to_polars_expression(base)

        return _LazyCell(self._expression.str.to_integer(base=base, strict=False))

    def to_lowercase(self) -> Cell[str | None]:
        return _LazyCell(self._expression.str.to_lowercase())

    def to_time(self, *, format: str | None = "iso") -> Cell[datetime.time | None]:
        if format == "iso":
            format = "%T%.f"  # noqa: A001
        elif format is not None:
            format = _convert_and_check_datetime_format(format, type_="time", used_for_parsing=True)  # noqa: A001

        return _LazyCell(self._expression.str.to_time(format=format, strict=False))

    def to_uppercase(self) -> Cell[str | None]:
        return _LazyCell(self._expression.str.to_uppercase())
