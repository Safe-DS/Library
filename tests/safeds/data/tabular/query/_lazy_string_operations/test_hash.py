import polars as pl
import pytest

from safeds.data.tabular.query import StringOperations
from safeds.data.tabular.query._lazy_string_operations import _LazyStringOperations


def test_should_be_deterministic() -> None:
    cell = _LazyStringOperations(pl.col("a"))
    assert hash(cell) == 8162512882156938440


@pytest.mark.parametrize(
    ("cell1", "cell2", "expected"),
    [
        (_LazyStringOperations(pl.col("a")), _LazyStringOperations(pl.col("a")), True),
        (_LazyStringOperations(pl.col("a")), _LazyStringOperations(pl.col("b")), False),
    ],
    ids=[
        "equal",
        "different",
    ],
)
def test_should_be_good_hash(cell1: StringOperations, cell2: StringOperations, expected: bool) -> None:
    assert (hash(cell1) == hash(cell2)) == expected
