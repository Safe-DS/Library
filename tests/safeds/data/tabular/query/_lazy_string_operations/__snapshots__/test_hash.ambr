# serializer version: 1
# name: TestContract.test_should_return_same_hash_in_different_processes[column]
  2158421768275916388
# ---
# name: TestContract.test_should_return_same_hash_in_different_processes[constant]
  1246368119593963133
# ---
